Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,789 @@
+        try await self.yield(contentsOf: CollectionOfOne(element))
+    }
+
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended or subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will return a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// - Note: Calling this function more than once has no effect.
+    @inlinable
+    public func finish() {
+        self._storage.finish(with: nil)
+    }
+
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended or subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will return a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// - Note: Calling this function more than once has no effect.
+    /// - Parameter failure: The failure indicating why the writer finished.
+    @inlinable
+    public func finish(with failure: Failure) {
+        self._storage.finish(with: failure)
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// The underlying sink of the ``NIOAsyncWriter``. This type allows to toggle the writability of the ``NIOAsyncWriter``.
+    public struct Sink {
+        @usableFromInline
+        /* private */ internal let _storage: Storage
+
+        @inlinable
+        init(storage: Storage) {
+            self._storage = storage
+        }
+
+        /// Sets the writability of the ``NIOAsyncWriter``.
+        ///
+        /// If the writer becomes writable again all suspended yields will be resumed and the produced elements will be forwarded via
+        /// the ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` method. If the writer becomes unwritable all
+        /// subsequent calls to ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` will suspend.
+        ///
+        /// - Parameter writability: The new writability of the ``NIOAsyncWriter``.
+        public func setWritability(to writability: Bool) {
+            self._storage.setWritability(to: writability)
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// This is the underlying storage of the writer. The goal of this is to synchronize the access to all state.
+    @usableFromInline
+    /* fileprivate */ internal final class Storage: @unchecked Sendable {
+        /// The lock that protects our state.
+        @usableFromInline
+        /* private */ internal let _lock = Lock()
+        /// The counter used to assign an ID to all our yields.
+        @usableFromInline
+        /* private */ internal let _yieldIDCounter = ManagedAtomic<UInt64>(0)
+        /// The state machine.
+        @usableFromInline
+        /* private */ internal var _stateMachine: StateMachine
+        /// The delegate.
+        @usableFromInline
+        /* private */ internal var _delegate: Delegate?
```

Comment:
I assume, holding the `Delegate`, or not holding the `Delegate` is part of the internal state. Should the delegate be referenced from the StateMachine? Then the stateMachine should return the Delegate when needed for some actions. This way, we should get rid of a bunch of `?`.

Reply:
Yeah happy to change this. Especially now that we are calling the delegate while holding the lock